packaging/_tokenizer.py

```diff
@@ -0,0 +1,153 @@
+        r"""
+            ('[^']*')
+            |
+            ("[^"]*")
+        """,
+        re.VERBOSE,
+    ),
+    "OP": r"===|==|~=|!=|<=|>=|<|>",
+    "VERSION": re.compile(Specifier._version_regex_str, re.VERBOSE | re.IGNORECASE),
+    "BOOLOP": r"or|and",
+    "IN": r"in",
+    "NOT": r"not",
+    "VARIABLE": re.compile(
+        r"""
+            python_version
+            |python_full_version
+            |os[._]name
+            |sys[._]platform
+            |platform_(release|system)
+            |platform[._](version|machine|python_implementation)
+            |python_implementation
+            |implementation_(name|version)
+            |extra
+        """,
+        re.VERBOSE,
+    ),
+    "URL_SPEC": "@ *[^ ]+",
+    "IDENTIFIER": r"([a-zA-Z0-9]|-|_|\.)+",
+}
+
+
+class Tokenizer:
+    """Stream of tokens for a LL(1) parser.
+
+    Provides methods to examine the next token to be read, and to read it
+    (advance to the next token).
+    """
+
+    next_token: Any
+
+    def __init__(
+        self, source: Sized, rules: Dict[Optional[str], object] = DEFAULT_RULES
+    ) -> None:
+        self.source = source
+        self.rules = {name: re.compile(pattern) for name, pattern in rules.items()}
+        self.next_token = None
+        self.generator = self._tokenize()
+        self.position = 0
+
+    def peek(self, *match_args: None, **match_kwargs: None) -> Any:
+        """Return the next token to be read"""
+        if not self.next_token:
+            self.next_token = next(self.generator)
+        return self.next_token
+
+    def match(self, *match_args: str, **match_kwargs: None) -> Any:
+        """Return True if the next token matches the given arguments"""
+        token = self.peek()
+        return token.matches(*match_args, **match_kwargs)
+
+    def expect(self, *match_args: str, **match_kwargs: None) -> Any:
+        """Raise SyntaxError if the next token doesn't match given arguments"""
+        token = self.peek()
+        if not token.matches(*match_args, **match_kwargs):
+            exp = " ".join(
+                v
+                for v in match_args
+                + tuple(f"{k}={v!r}" for k, v in match_kwargs.items())
+                if v
+            )
+            raise self.raise_syntax_error(f"Expected {exp}")
+        return token
+
+    def read(self, *match_args: str, **match_kwargs: None) -> Any:
+        """Return the next token and advance to the next token
+
+        Raise SyntaxError if the token doesn't match.
+        """
+        result = self.expect(*match_args, **match_kwargs)
+        self.next_token = None
+        return result
+
+    def try_read(self, *match_args: str, **match_kwargs: None) -> Any:
+        """read() if the next token matches the given arguments
+
+        Do nothing if it does not match.
+        """
+        if self.match(*match_args, **match_kwargs):
+            return self.read()
+
+    def raise_syntax_error(self, message: str = "Invalid marker") -> Any:
```

Comment:
The message should likely not have a default value, and instead be a required kwarg.